### DISCUSSION TITLE
Suggestion - TinyGo organization by Microcontrollers

### DISCUSSION THREAD

**Discussion by Gustavomurta** [tinygo-org/tinygo#3959] [Ideas]
I don't have much experience in Github, but I think that if the machine packages were subdivided by microcontrollers, it would be easier for users. For example: 

https://github.com/tinygo-org/tinygo/tree/release/src/machine

Generic
Atmega AVR 
Atmega SAMD
RP2040
STM32

**Comment by aykevl** [comment]
How do you suggest this would be implemented, exactly? For example, do you propose to change the machine package API?

**Reply by Gustavomurta** [reply]
If it takes a lot of work to organize, please disregard my suggestion. My intention is to collaborate.

**Reply by soypat** [reply]
I don't think the problem is organizing it, but rather go requires each directory to be a package that is importable via it's name. To maintain the machine HAL abstraction we'd be unable to use the go compiler as is because importing need to import `machine/rp2040` as `machine "machine/rp2040"` which would break when switching to atmega avr targets... unless we were to modify the go compiler to interpret the `"machine" import as a subdirectory depending on the `target` flag... but this seems like it gives us less than what it takes to implement and the complexity that arises.
